daemon: Add tofqdns-pre-cache DNS data preload

raybejjani opened the pull request:
When restarting cilium-agent, DNS data may be lost, particularly when
upgrading from earlier cilium versions without a DNS restore feature. We now allow preloading the DNS cache data for use by L3 toFQDN rules in cilium-agent.
Separately, it may also be useful to always ensure some IPs are
available as a minimum set for a DNS name. This is redundant with
toCIDR, however, so it is mostly a convenience for toFQDNs situations.

This can be folded into https://github.com/cilium/cilium/pull/6886 instead, if that's preferred. I'll post a similar message there. This option removes some of the complexity around the plan proposed in https://github.com/cilium/cilium/pull/6886. From a user perspective, it side-steps determining when to remove the generated bridge-CNP as the IPs behave like they would in the normal cache (and better accounts for upgrades from 1.3 which did have a cache). From my perspective, it allows me to avoid developing a script/tool that needs to exec on each cilium pod and then create a global CNP (or print it). This proved a little more involved than I expected. Instead, a preflight container can do the work (see below).


The upgrade plan with this option would be:
- the cilium preflight image is updated to run a new script. It runs on all nodes, adjacent to cilium-agent.
  - mount /var/run/cilium
  - Use the cilium API to `cilium policy get` directly with cilium.sock
  - Create a json file of pretend DNS lookups in a file under `/var/run/cilium/dns-precache-upgrade-1.4.json`. The expiration time is absolute, and it will expire out like any other DNS data.
  - The pod posts an error if this fails, making it clear the upgrade cannot proceed
- User updates the cilium configmap/daemonset to add `--tofqdns-pre-cache /var/run/cilium/dns-precache-upgrade-1.4.json` (or the env var equivalent).
- Switch to the cilium 1.4 image. Cilium 1.4 preloads this data into the cache at the same stage where it restores the endpoint caches. Policies received from K8s will be filled in with this data.
- The file can be deleted or not. The cache will expire the pre-cache data normally (in fact, it will be replaced by more recent lookups from the poller or DNS proxy).

An open question is whether we want to switch the DNSCache json format to match the one in the `cilium fqdn cache list` format. The fields values are the same, but they are named differently. This would break restore from other 1.4-RC images and master, but not from other versions because this feature hasn't been released. I've not included this here in case we treat the RCs are releases.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6979)
<!-- Reviewable:end -->

raybejjani committed:
daemon: Add tofqdns-pre-cache DNS data preload

When restarting cilium-agent, DNS data may be lost, particularly when
upgrading from earlier cilium versions without a DNS restore feature. We
now allow preloading the DNS cache data for use by L3 toFQDN rules in
cilium-agent.
Separately, it may also be useful to always ensure some IPs are
available as a minimum set for a DNS name. This is redundant with
toCIDR, however, so it is mostly a convenience for toFQDNs situations.

Signed-off-by: Ray Bejjani <ray@covalent.io>

aanm commented on daemon/daemon_main.go:
```diff
@@ -670,6 +670,9 @@ func init() {
 	flags.Int(option.ToFQDNsMaxIPsPerHost, defaults.ToFQDNsMaxIPsPerHost, "Maximum number of IPs to maintain per FQDN name for each endpoint")
 	option.BindEnv(option.ToFQDNsMaxIPsPerHost)
 
+	flags.StringVar(&option.Config.ToFQDNsPreCache, option.ToFQDNsPreCache, defaults.ToFQDNsPreCache, "DNS cache data at this path is preloaded on agent startup")
```
ah, so you are the one that keeps adding the StringVar thingy. You can use the `flags.String(` which doesn't need you to pass `option.Config.ToFQDNsPreCache`. Jut make sure this is assigned like all flags under `pkg/config/config.go` on lines 990s

raybejjani replied:
oh, I just copy-pasted another thing... I'll switch it :):

tgraf commented on daemon/fqdn.go:
```diff
@@ -159,6 +160,20 @@ func (d *Daemon) bootstrapFQDN(restoredEndpoints *endpointRestoreState) (err err
+			// than to enter crash backoffs when this file is broken.
+		}
+
+		fqdn.DefaultDNSCache.UpdateFromCache(precache, nil)
```
Is it safe to use `precache` here even if `err != nil` above?

raybejjani replied:
It is because UpdateFromCache happens to no-op when the source-cache is nil. I'll make this the else case of the error handler since that is clearer (also, I wasn't paying attention, definitely not intended). Good catch

raybejjani committed:
fqdn: Coordinate DNSCache JSON names with API models

We have a JSON representation for cacheEntry objects in the cache, used
when serializing state, and a different one returned by the cilium API
for `/fqdn/cache` responses. These are almost identical, and making them
the same allows using `cilium fqdn cache list -o json` to generate files
for use with `--tofqdns-pre-cache`. It also means the API documentation
applies for the format of this file.

Signed-off-by: Ray Bejjani <ray@covalent.io>
